File: skypydb-js/src/embeddings/mixins/get_embedding.ts
import type { EmbeddingVector } from "../../types";

type EmbeddingProviderLike = {
  _get_embedding?: (text: string) => Promise<EmbeddingVector>;
};

export async function get_embedding(
  provider: unknown,
  text: string
): Promise<EmbeddingVector> {
  const candidate = provider as EmbeddingProviderLike;

  if (typeof candidate._get_embedding === "function") {
    return candidate._get_embedding(text);
  }

  throw new Error(
    `${(candidate as { constructor?: { name?: string } }).constructor?.name ?? "Provider"} must implement _get_embedding.`
  );
}

File: skypydb/embeddings/core/get_embedding.py
"""Get embedding provider factory helpers."""

from typing import List


def get_embedding(
    self,
    text: str,
) -> List[float]:
    """Get embedding for a single text."""

    embedding_impl = getattr(self, "_get_embedding", None)
    if not callable(embedding_impl):
        raise NotImplementedError(
            f"{self.__class__.__name__} must implement `_get_embedding`."
        )
    return embedding_impl(text)